### PR TITLE
Add nmrec dependency

### PR DESCRIPTION
**Pull request by seth127**
_Work in progress... adds nmrec as a dependency and refactors some internal functions to use it, in place of legacy code that primarily relied on regex._

### CODE REVIEW

**Comment by seth127** [comment]
@kyleam do you have an opinion about whether it would be easier/smoother to merge this and make a bbr release before or after nmrec gets added to MPN?

In other words:
- nmrec will be on the Sept 2023 snapshot
- should we make a release with this stuff and put it on the Sept 2023 snapshot, or is it easier to wait until after that snapshot?

**Comment by kyleam** [comment]
> is it easier to wait until after that snapshot?

Unless you think there's any urgency for this rework to land on MPN (I'm not aware of any), I'd prefer to wait until nmrec makes it to snapshot.  We've got a good number of moving parts already with the snapshot and mrg releases, and waiting would let us finalize things (e.g., the drone config) without requiring post-release followups.

**Comment by seth127** [comment]
Ok that makes sense to me @kyleam . Let's plan to revisit this in early October, once that MPN snapshot is out.

**Comment by kyleam** [comment]
The snapshot is out; I've adjusted the Drone config to drop the custom nmrec installation in all spots aside from the oldest build, which still requires it (2e1edb91).

https://github-drone.metrumrg.com/metrumresearchgroup/bbr/3549

**Comment by seth127** [comment]
@kyleam @barrettk what's the status of this branch at this point? I know it was serving as a dev branch for awhile... are there still pending PRs coming into here, or should we go ahead and merge it?

Related, @barrettk is actively working on #582 , so we could until that is done to merge all of this. That will likely be a little while though.

I'm fine with whatever y'all want to do, I just wanted to make sure this wasn't just loitering at this point. Thanks.

**Comment by kyleam** [comment]
@seth127, once nmrec hit MPN, I pushed an update for the CI (2e1edb9), so it's ready to merge in my view. We no longer need a branch to merge nmrec PRs now that that nmrec is on MPN.

(Note that there's the unmerged gh-605, but [my opinion](https://github.com/metrumresearchgroup/bbr/pull/605#issuecomment-1660822495) is that that can be closed without merge.)

> Related, @barrettk is actively working on #582 [...]

I don't think that we should wait for that to come in.

> I'm fine with whatever y'all want to do, I just wanted to make sure this wasn't just loitering at this point.

As the opener of the PR, I was expecting you to switch it to "ready for review" and guide it to completion (probably just requesting a review and hitting merge at this point, though things may pop up).  Would you like someone to take over that?

File: .drone.yml
Comment (barrettk) [inline .drone.yml:105]: I think we want to change these lines before merging FYI

**Comment by seth127** [comment]
Thanks for the updates.

> As the opener of the PR, I was expecting you to switch it to "ready for review" and guide it to completion (probably just requesting a review and hitting merge at this point, though things may pop up). Would you like someone to take over that?

Yes, fair point. Although I would be happy if you don't mind taking it over @kyleam . I think you've reviewed most (all?) of the PRs that came _into_ this branch, so hopefully it will be pretty quick look this one over. Let me know if I'm missing any nuance there and we can discuss a different plan, if needed.

File: .drone.yml
Comment (kyleam) [reply .drone.yml:105]: > I think we want to change these lines before merging FYI

What is your suggested change?

Comment (kyleam) [reply .drone.yml:105]: As the comment on line 103 says, nmrec isn't available on this image's (`r-dev-ci-mpn-4.0:2020-06-08`) snapshot, so we need to keep the custom install.  We could update the custom command to install nmrec from the 2023-09-19 snapshot instead of the Git repo (or equivalently just make this use the v0.2.0 tag from the nmrec repo).  

I could see the argument that pinning the nmrec version is more inline with the "test oldest" goal of this build.  Is that the reason for your suggestion?

Comment (kyleam) [reply .drone.yml]: > I could see the argument that pinning the nmrec version is more inline with the "test oldest" goal of this build

a92c9bf4 pins the nmrec version.

**Review by kyleam — APPROVED** [review]
The core changes of this series have been reviewed elsewhere: 600, 603, 604

Here are the non-merge commits directly to this branch:

 * 5a1b11b7 (add nmrec dependency to drone)
 * 1734b8d6 (add nmrec to description)
 * 8f8e40de (document unquote_filename)
 * 2e1edb91 (ci: drop now unneeded nmrec installations)
 * a92c9bf4 (ci: pin oldest pipeline to nmrec 0.2.0)

Those are non-R changes: tweak the Drone configuration, update `DESCRIPTION` to include nmrec as dependency, and regenerate missing docs.  Those all look fine to me.

---

At least on my end, GitHub isn't showing the Drone builds as completed.  Here they are (both green):

 * https://github-drone.metrumrg.com/metrumresearchgroup/bbr/3555
 * https://github-drone.metrumrg.com/metrumresearchgroup/bbr/3556

**Comment by kyleam** [comment]
@barrettk If all looks good to you, please do the honors of merging to main.

File: .drone.yml
Comment (barrettk) [reply .drone.yml]: >I could see the argument that pinning the nmrec version is more inline with the "test oldest" goal of this build. Is that the reason for your suggestion?

I thought about that too, but mainly it was just to reduce some of the variability in what version of `nmrec` we pull in drone. In the event `nmrec` and `bbr` were undergoing larger refactors, I figured it would be good to have a more stable version of `nmrec` rather than whatever was on `main` at the time. Pointing to a specific version definitely takes care of that and IMO is actually better than pointing to an MPN snapshot anyways (so we can track `nmrec` and `bbi` versions similarly).

**Comment by barrettk** [comment]
>At least on my end, GitHub isn't showing the Drone builds as completed. Here they are (both green)

Yeah that is the same on my end as well

File: .drone.yml
Comment (kyleam) [reply .drone.yml]: > mainly it was just to reduce some of the variability in what version of `nmrec` we pull in drone.

Having the nmrec version be different across the pipelines is fine, and in many cases desirable because it increases what you're testing.  On average nmrec's main should be more stable than a release because it has fixes, but in the event that it does break bbr compatibility, we want to know because it shouldn't.

In any case, for the oldest build using an old nmrec for consistency as a92c9bf works for me, so we're set for the purposes of this PR.

**Comment by kyleam** [comment]
The good old off/on approach didn't work to get the status checks to show up.  (Given it's not urgent to get this in, I'm okay holding this up to see if we can resolve it; however, I'm also fine with admin-bypassing this and waiting to see whether it comes up again.)

**Comment by seth127** [comment]
Per [this comment above](https://github.com/metrumresearchgroup/bbr/pull/606#pullrequestreview-1668573637) everything is good, we just have a glitch in the Drone build showing up. I am merging with my admin privileges to avoid having to dig any deeper into the Drone issue. We'll dig deeper on it if this keeps happening on other PRs.

**Pull request merged**
